crates/sui-core/src/quorum_driver/mod.rs

```diff
@@ -221,6 +274,163 @@ where
         self.validators.store(new_validators);
         Ok(())
     }
+
+    // TODO currently this function is not epoch-boundary-safe. We need to make it so.
+    /// Returns Ok(None) if the no conflicting transaction was retried.
+    /// Returns Ok(Some((tx_digest, true))) if one conflicting transaction was retried and succeeded,
+    /// Some((tx_digest, false)) otherwise.
+    /// Returns Error on unexpected errors.
+    #[allow(clippy::type_complexity)]
+    async fn attempt_conflicting_transactions_maybe(
+        &self,
+        good_stake: StakeUnit,
+        conflicting_tx_digests: &BTreeMap<
+            TransactionDigest,
+            (Vec<(AuthorityName, ObjectRef)>, StakeUnit),
+        >,
+        original_tx_digest: &TransactionDigest,
+    ) -> SuiResult<Option<(TransactionDigest, bool)>> {
+        let validity = self.validators.load().committee.validity_threshold();
+        // if we have >= f+1 good stake on the current transaction, no point in retrying conflicting ones
+        if good_stake >= validity {
+            return Ok(None);
+        }
+
+        let mut conflicting_tx_digests = Vec::from_iter(conflicting_tx_digests.iter());
+        conflicting_tx_digests.sort_by(|lhs, rhs| rhs.1 .1.cmp(&lhs.1 .1));
+        if conflicting_tx_digests.is_empty() {
+            error!("This path in unreachable with an emtpy conflicting_tx_digests.");
+            return Ok(None);
+        }
+
+        // we checked emptiness above, safe to unwrap.
+        let (tx_digest, (validators, total_stake)) = conflicting_tx_digests.get(0).unwrap();
+
+        // To be more conservative and try not to actually cause full equivocation,
+        // we only retry a transaction when at least f+1 validators claims this tx locks objects
+        if *total_stake < validity {
+            return Ok(None);
+        }
+        info!(
+            ?tx_digest,
+            ?total_stake,
+            ?original_tx_digest,
+            "retrying conflicting tx."
+        );
+        let is_tx_executed = self
+            .attempt_one_conflicting_transaction(
+                tx_digest,
+                original_tx_digest,
+                validators
+                    .iter()
+                    .map(|(name, _obj_ref)| name)
+                    .collect::<Vec<_>>(),
+            )
+            .await?;
+
+        Ok(Some((**tx_digest, is_tx_executed)))
+    }
+
+    /// Returns Some(true) if the conflicting transaction is executed successfully
+    /// (or already executed), or Some(false) if it did not.
+    async fn attempt_one_conflicting_transaction(
```

Comment:
I recommend we simplify this function by:
1. Move the part that gets a raw transaction from a list of validators to AuthorityAggregator. The function API should take a digest, a list of authority names and a timeout, and return a `VerifiedTransaction`.
2. In the above function, we always return the raw transaction. If a validator returns a cert, we take the transaction out of the cert; if a validator returns a signed tx, we take the transaction out of the signed tx. Either way the API returns a raw transaction
3. Then here we just ask all validators to process the tx.

Reply:
sounds good on moving it to auth-agg. 
But why do you think it's better to always return `VerifiedTransaction`? If there's a valid Cert we can skip the `handle_transaction` step and directly do `handle_certificate`

Reply:
That's fair. It's an optimization, but perhaps it's an easy one to do. I guess you could return an Either<VerifiedTransaction, VerifiedCertificate>. Then depends on the result do different things.